src/KubernetesClient/Watch.cs

```diff
@@ -0,0 +1,482 @@
+        }
+
+        /// <summary>Initializes a new <see cref="WatchReader{T}"/> from a <see cref="Stream"/> containing a watch stream.</summary>
+        public WatchReader(Stream stream) => CreateReader(stream ?? throw new ArgumentNullException(nameof(stream)));
+
+        /// <summary>Indicates whether the request failed and so the <see cref="Response"/> does not contain a valid watch stream.</summary>
+        public bool IsError => Response != null && (int)Response.StatusCode >= 400;
+
+        /// <summary>Gets the underlying <see cref="HttpResponseMessage"/>, or null if the watch reader was initialized from a <see cref="Stream"/>.</summary>
+        public HttpResponseMessage Response { get; }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            streamReader?.Dispose();
+            Response?.Dispose();
+        }
+
+        /// <summary>Reads the next <see cref="WatchEvent{T}"/> from the stream, or returns null if the stream was closed or
+        /// <see cref="IsError"/> is true.
+        /// </summary>
+        public async Task<WatchEvent<T>> ReadAsync(CancellationToken cancelToken = default)
+        {
+            if (IsError) return null;
+            cancelToken.ThrowIfCancellationRequested();
+            if (reader == null) CreateReader(await Response.Content.ReadAsStreamAsync().ConfigureAwait(false));
+            try
+            {
+                streamReader.CancelToken = cancelToken;
+#if NETCOREAPP2_1
+                if (!await reader.ReadAsync(cancelToken).ConfigureAwait(false)) return null; // wait for the next event
+#else
+                if (!reader.Read()) return null; // wait for the next event
+#endif
+            }
+            catch (IOException ex) when (ex.InnerException is WebException wex && wex.Status == WebExceptionStatus.RequestCanceled)
+            {
+                throw new OperationCanceledException("The watch read was canceled.", ex.InnerException, cancelToken);
+            }
+
+            // deserialize the watch event
+            var e = new WatchEvent<T>();
+            bool gotType = false;
+            while (true)
+            {
+                reader.Read(); // move to the next property, if any
+                if (reader.TokenType != JsonToken.PropertyName) break;
+                string name = (string)reader.Value;
+                reader.Read(); // move to the property value
+                if (name == "type")
+                {
+                    e.Type = (WatchEventType)Enum.Parse(typeof(WatchEventType), (string)reader.Value, true);
+                    gotType = true;
+                }
+                else if (name == "object")
+                {
+                    if (!gotType) throw new JsonSerializationException("Expected type property before object.");
+                    if (e.Type != WatchEventType.Error) e.Object = serializer.Deserialize<T>(reader);
```

Comment:
If I have understood correctly: in the (not unlikely) case an event spans the 4 KiB buffer, then `Deserialize` will result in blocking IO.  Also, I believe there are pathological cases where a disproportionate amount of IO becomes blocking (e.g. burst of events slightly smaller than some multiple of the buffer size).

It would be great if this were made async all the way.

Reply:
The CancellableStreamReader can cancel even normal blocking IO by closing the stream when the token is canceled, though I'm unsure of what exception will be thrown if any in that case.

Reply:
Cancellation is not my concern, it's that blocking IO will result in blocked thread pool threads. Admittedly, threads may only be blocked for short periods, but in a system with many watchers it may cause the number of threads to grow large.

Can it not be made async all the way?

Reply:
I think that is sufficient, and furthermore I don't see an async version of Deserialize or an equivalent, but let me know if I'm missing something.

Reply:
I can make the calls to Read async. I don't know about Deserialize.

Reply:
Here's a potential implementation for `!NET452`!
```csharp
try
{
  await reader.ReadAsync(cancelToken).ConfigureAwait(false);
  // Load.
  var json = await JObject.LoadAsync(reader, cancelToken).ConfigureAwait(false);
  var type = (string)json.Property("type");
  var obj = json.Property("object")?.Value;
  if (type == null || obj == null) throw new JsonSerializationException("The stream does not appear to contain watch events.");
  // Deserialize.
  var e = new WatchEvent<T>();
  e.Type = (WatchEventType)Enum.Parse(typeof(WatchEventType), type, true);
  if (e.Type != WatchEventType.Error) e.Object = obj.ToObject<T>(serializer);
  else e.Error = obj.ToObject<V1Status>(serializer);
  return e;
}
catch (IOException ex) when (ex.InnerException is WebException wex && wex.Status == WebExceptionStatus.RequestCanceled)
{
    throw new OperationCanceledException("The watch read was canceled.", ex.InnerException, cancelToken);
}
```
**NB** it was hacking this that identified the `charRead` issues.

Reply:
Hmm. I'm glad to know it's possible. I'm not sure which is better, but I'm not fond of how it goes from JSON -> JObject -> T (allocating a bunch of memory for the object tree) rather than just going from JSON -> T, especially since in practice the server isn't going to send half an event and then stop (so it shouldn't be blocked on IO for any significant length of time, which was the same rationale I had for the synchronous calls to Read()).

Reply:
it is really a headache if we cannot do all the way async
the event json was separate by line ending inside the stream
another way is to create a buf for current line
then fill the buf asyncly.

due to json cannot be stream-parsed, you have to buf entire line to read json

Reply:
What is the headache, by the way? The most important place to avoid blocking is while waiting for an event, since that can take an arbitrary amount of time. Once the server decides to send an event, however, it sends it in its entirety. (The server doesn't send half an event and then stop.) Thus, while it's crucial to have the ability to cancel the wait for an event and async helps avoid blocking a thread there, it's not so important when parsing an event that is already being sent over the network.